packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/queue/queue.ts

```diff
@@ -0,0 +1,389 @@
+  private environment: { [key: string]: string };
+
+  constructor(props: ContainerMutatingProps) {
+    super();
+    this.environment = props.environment;
+  }
+
+  public mutateContainerDefinition(props: ecs.ContainerDefinitionOptions): ecs.ContainerDefinitionOptions {
+    return {
+      ...props,
+
+      environment: { ...(props.environment || {}), ...this.environment },
+    } as ecs.ContainerDefinitionOptions;
+  }
+}
+
+/**
+ * This extension creates a default `eventsQueue` for the service (if not provided) and accepts a list of objects of
+ * type `ISubscribable` that the `eventsQueue` subscribes to. It creates the subscriptions and sets up permissions
+ * for the service to consume messages from the SQS Queues.
+ *
+ * It also configures auto scaling for the service by scaling up or down to maintain an acceptable queue latency by
+ * tracking the accpetable backlog per task.
+ *
+ * The default queue for this service can be accessed using the getter `<extension>.eventsQueue`.
+ */
+export class QueueExtension extends ServiceExtension {
+  private _eventsQueue!: sqs.IQueue;
+
+  private _eventsQueueDelay?: QueueAutoScalingOptions;
+
+  private subscriptionQueues = new Set<SubscriptionQueue>();
+
+  private environment: { [key: string]: string } = {};
+
+  private props?: QueueExtensionProps;
+
+  public logGroup?: logs.ILogGroup;
+
+  constructor(props?: QueueExtensionProps) {
+    super('queue');
+
+    this.props = props;
+  }
+
+  /**
+   * This hook creates (if required) and sets the default queue `eventsQueue`. It also sets up the subscriptions for
+   * the provided `ISubscribable` objects.
+   *
+   * @param service The parent service which this extension has been added to
+   * @param scope The scope that this extension should create resources in
+   */
+  public prehook(service: Service, scope: Construct) {
+    this.parentService = service;
+    this.scope = scope;
+
+    let eventsQueue = this.props?.eventsQueue;
+    if (!eventsQueue) {
+      const deadLetterQueue = new sqs.Queue(this.scope, 'EventsDeadLetterQueue', {
+        retentionPeriod: cdk.Duration.days(14),
+      });
+
+      eventsQueue = new sqs.Queue(this.scope, 'EventsQueue', {
+        deadLetterQueue: {
+          queue: deadLetterQueue,
+          maxReceiveCount: 3,
+        },
+      });
+    }
+    this._eventsQueue = eventsQueue;
+    this._eventsQueueDelay = this.props?.scaleOnLatency;
+
+    this.environment[`${this.parentService.id.toUpperCase()}_QUEUE_URI`] = this._eventsQueue.queueUrl;
+
+    if (this.props?.subscriptions) {
+      for (const subs of this.props.subscriptions) {
+        const subsQueue = subs.subscribe(this);
+        if (subsQueue.queue !== this._eventsQueue) {
+          if (subsQueue.scaleOnLatency && !this._eventsQueueDelay) {
+            throw Error(`Cannot configure auto scaling only for the subscription queue of service '${this.parentService.id}'. Need to specify 'eventsQueueDelay'.`);
```

Comment:
Why is this the case? do we need to also update the error message?
```suggestion
            throw Error(`Cannot configure auto scaling only for the subscription queue of service '${this.parentService.id}'. Need to specify 'scaleOnLatency'.`);
```

Reply:
`scaleOnLatency` provided in the `QueueExtension` will be used for calculating the acceptable backlog for all queues in the service unless the user specifies a different input in the the `TopicSubscription`. If provided, the new input will then override the default `scaleOnLatency` for the topic-specific queue.

This is not very clear from the error message. Updated!! But hope it makes sense to do this 🙈